### ISSUE TITLE
Add transaction proposal capability to the light client FFI

### ISSUE THREAD

**Issue by nuttycom** [Electric-Coin-Company/zcash-light-client-ffi#86]
In https://github.com/zcash-hackworks/zcash-light-client-ffi/blob/main/rust/src/lib.rs, the methods `zcashlc_create_to_address` and `zcashlc_shield_funds` should be refactored to call [propose_transfer](https://github.com/zcash/librustzcash/blob/main/zcash_client_backend/src/data_api/wallet.rs#L326) and [propose_shielding] (https://github.com/zcash/librustzcash/blob/main/zcash_client_backend/src/data_api/wallet.rs#L364) respectively, and then use [create_proposed_transaction](https://github.com/zcash/librustzcash/blob/main/zcash_client_backend/src/data_api/wallet.rs#L405) in each case to actually perform the transaction creation.

The `Proposal` so created should be exposed over the FFI as a boxed pointer, so that the SDK can create a proposal, query it over the FFI to obtain the fee information from it, and then provide that proposal back over the FFI when calling `create_proposed_transaction` which should consume the box and free the associated memory.

**Comment by softminus** [comment]
so @nuttycom had suggested:
> So, here's what I imagine:
> A new method,` zcashlc_propose_transfer_internal` which takes the same arguments as `zcashlc_create_to_address` and is basically the entirety of the `zcashlc_create_to_address` method up to where it calls `spend`. Instead of calling `spend`, it calls `propose_transfer` and returns a `Proposal`.
> Then, `zcashlc_create_to_address` and a new method, `zcashlc_propose_transfer` both call `zcashlc_propose_transfer_internal`. In the case of the latter, it calls `Box::into_raw(Box::new(proposal))` and returns the pointer to the box; in the case of the former, it simply calls `create_proposed_transaction` immediately.

I tried to get this to work, but I ran into issues getting the `zcashlc_propose_transfer_internal` return type working for both the zip317 and fixed fees, since the type of the Proposal that gets returned depends on the fee type. I understand that it will get `Box`'d and so the two types don't need to match up, but I haven't found a way to write something like `Box<Proposal<dyn FeeRule, NoteID>> ` for the return type -- which would work for zip317 fees and fixed fees at all.

I did get it to compile with handling only zip317 fees and a corresponding return value type of `*mut Proposal<zcash_primitives::transaction::fees::zip317::FeeRule, NoteId>`, but I don't know how I can write a return type (and construct a value for it in both cases) for both fixed and zip317 fees.

This is how the types work for propose_transfer-- but I don't understand how we can do the equivalent of that here -- with how the `InputsT` and `DbT` types create the type for the `Proposal<InputsT::FeeRule, DbT::NoteRef>` return value:
```
pub fn propose_transfer<DbT, ParamsT, InputsT>(
    wallet_db: &mut DbT,
    params: &ParamsT,
    spend_from_account: AccountId,
    input_selector: &InputsT,
    request: zip321::TransactionRequest,
    min_confirmations: u32,
) -> Result<
    Proposal<InputsT::FeeRule, DbT::NoteRef>,
    Error<DbT::Error, InputsT::Error, <InputsT::FeeRule as FeeRule>::Error, DbT::NoteRef>,
>
where
    DbT: WalletWrite,
    DbT::NoteRef: Copy + Eq + Ord,
    ParamsT: consensus::Parameters + Clone,
    InputsT: InputSelector<DataSource = DbT>,
{ 
```


i tried to write `mut<Proposal<dyn FeeRule, NoteID>>` and we get this error:
```
error[E0191]: the value of the associated type `Error` (from trait `zcash_primitives::transaction::fees::FeeRule`) must be specified
    --> src/lib.rs:2213:24
     |
2213 | ) -> *mut Proposal<dyn FeeRule, NoteId>   {
     |                        ^^^^^^^ help: specify the associated type: `FeeRule<Error = Type>
```
and it's unclear what value of Error the FeeRule trait takes (if that's even the right trait to be using)

and if we feed it the type of `Error` that `propose_transfer` returns (again, not sure if this is the right error type) we get:

```
error[E0038]: the trait `zcash_primitives::transaction::fees::FeeRule` cannot be made into an object
    --> src/lib.rs:2213:20
     |
2213 | ...al<dyn FeeRule<Error= zcash_client_backend::data_api::error::Error<SqliteClientError, GreedyInputSelectorError<FeeError, NoteId>, FeeError, NoteId>>, N...
     |       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `zcash_primitives::transaction::fees::FeeRule` cannot be made into an object
     |
note: for a trait to be "object safe" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
    --> /Users/sasha_work/.cargo/registry/src/github.com-1ecc6299db9ec823/zcash_primitives-0.10.0/src/transaction/fees.rs:24:8
     |
24   |     fn fee_required<P: consensus::Parameters>(
     |        ^^^^^^^^^^^^ the trait cannot be made into an object because method `fee_required` has generic type parameters
```

## duplicate everything?
Duplicating everything for zip317 and fixed-fee `Proposals` would mean mean a duplicate of all the methods if we can't find a way to write a type for zip317 and fixed-fee `Proposals` that can be passed across the FFI -- which seems incredibly inconvenient and weird.

## serialize it!
@nuttycom suggested we write code to serialize/deserialize the `Proposal` into a protobuf in `zcash_client_backend`, and when we return a `Proposal` across an FFI boundary, we return it in serialized form (and in turn, expect serialized `Proposal`s when we get called over the FFI).

This has the advantage of 1) solving the type complexity, 2) allowing users on the other side of the FFI to inspect the `Proposal` without us needing to write FFI methods to extract elements of the `Proposal`.

When we're fed a serialized `Proposal`, we'll deserialize it, and figure out if it's a zip317 or a fixed-fee `Proposal` and appropriately handle it

**Comment by softminus** [comment]
(I'm currently working on a protobuf for the `Proposal`s)

**Comment by softminus** [comment]
so far I have this: https://gist.github.com/softminus/0b9ac3b67eafcdca3f605b7fdf575793

**Comment by softminus** [comment]
Here's the protobuf definition for the `Proposal` type I have so far https://gist.github.com/softminus/0b9ac3b67eafcdca3f605b7fdf575793

**Comment by str4d** [comment]
Closed by #111, merged into the `ffi-0.5.0` feature branch.
